plugins/woocommerce-admin/client/marketing/hooks/test/useCampaigns.test.ts

```diff
@@ -0,0 +1,230 @@
+/**
+ * External dependencies
+ */
+import { renderHook } from '@testing-library/react-hooks/dom';
+import { waitFor } from '@testing-library/react';
+import { dispatch } from '@wordpress/data';
+import 'whatwg-fetch'; /* eslint-disable-line import/no-unresolved */ /* To make sure Response is available */
+
+/**
+ * Internal dependencies
+ */
+import { useCampaigns } from '../useCampaigns';
+import { useRegisteredChannels } from '../useRegisteredChannels';
+import { STORE_KEY } from '../../data-multichannel/constants';
+import { RegisteredChannel } from '../../types';
+import { Campaign as APICampaign } from '../../data-multichannel/types';
+import '../../data-multichannel'; // To ensure the store is registered
+
+type Channel = Pick< RegisteredChannel, 'slug' | 'title' | 'icon' >;
+
+jest.mock( '@wordpress/api-fetch', () =>
+	jest.fn( ( { path } ) => {
+		const total = 9;
+
+		const params = new URLSearchParams( path.replace( /^[^?]*/, '' ) );
+		const page = Number( params.get( 'page' ) );
+		const perPage = Number( params.get( 'per_page' ) );
+
+		if ( ! Number.isInteger( page ) || ! Number.isInteger( perPage ) ) {
+			return Promise.reject(
+				new Response( '{"message": "invalid query"}', { status: 400 } )
+			);
+		}
+
+		const length = Math.min( perPage, total - ( page - 1 ) * perPage );
+
+		const campaigns: Array< APICampaign > = Array.from( { length } ).map(
+			( _, index ) => {
+				const id = `${ page }_${ index + 1 }`;
+				return {
+					id,
+					channel: 'extension-foo',
+					title: `Campaign ${ id }`,
+					manage_url: `https://test/extension-foo?path=setup&id=${ id }`,
+					cost: {
+						value: ( ( page * perPage + index ) * 0.25 ).toString(),
+						currency: 'USD',
+					},
+				};
+			}
+		);
+
+		// For testing fallbacks when data fields are not available
+		if ( campaigns[ 2 ] ) {
+			campaigns[ 2 ].channel = 'intentional-mismatch-channel';
+			campaigns[ 2 ].cost = null;
+		}
+
+		return Promise.resolve(
+			new Response( JSON.stringify( campaigns ), {
+				headers: new Headers( { 'x-wp-total': total.toString() } ),
+			} )
+		);
+	} )
+);
```

Comment:
❔ I have some reservations / hesitance about mocking "@wordpress/api-fetch" here, because this means that we are also indirectly testing wp-data redux things here, i.e. the selector, resolver, reducer etc, in this `useCampaigns.test.js` file.

My thinking is that `useCampaigns.test.js` should be unit test on `useCampaigns`, and we would be mocking `useSelect`. But what we have here is more like integration test with wp-data, which is fine too, but just a different expectation. 😄 

So I think it's fine to keep it this way. Just want to express my expectation and opinion. If you can share your experience or thinking here, that would be great for us to learn.

Reply:
Got it. Thank you for the explanation! 🙂 👍